src/huggingface_hub/__init__.py

```diff
@@ -16,70 +16,174 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# ***********
+# vendored from https://github.com/scientific-python/lazy_loader
+import importlib
+import importlib.util
+import inspect
+import os
+import sys
+import types
+import warnings
+
+
+class LazyImportWarning(Warning):
+    pass
+
+
+def attach(package_name, submodules=None, submod_attrs=None):
+    """Attach lazily loaded submodules, functions, or other attributes.
+
+    Typically, modules import submodules and attributes as follows::
+
+      import mysubmodule
+      import anothersubmodule
+
+      from .foo import someattr
+
+    The idea is to replace a package's `__getattr__`, `__dir__`, and
+    `__all__`, such that all imports work exactly the way they would
+    with normal imports, except that the import occurs upon first use.
+
+    The typical way to call this function, replacing the above imports, is::
+
+      __getattr__, __dir__, __all__ = lazy.attach(
+        __name__,
+        ['mysubmodule', 'anothersubmodule'],
+        {'foo': ['someattr']}
+      )
+
+    This functionality requires Python 3.7 or higher.
+
+    Parameters
+    ----------
+    package_name : str
+        Typically use ``__name__``.
+    submodules : set
+        List of submodules to attach.
+    submod_attrs : dict
+        Dictionary of submodule -> list of attributes / functions.
+        These attributes are imported as they are used.
+
+    Returns
+    -------
+    __getattr__, __dir__, __all__
```

Comment:
We've also got a package at https://pypi.org/project/lazy_loader/ if you prefer that alternative.  Either way, you can keep up to date with improvements at https://github.com/scientific-python/lazy_loader

Reply:
Since the package is still experimental, I think it'd make sense for us to not have a dependency on it since the API might change and that would mean strict dependency constraints which would make it hard for `pip` to resolve dependencies.